### PR TITLE
[MISC] Mammogram screening model

### PR DESCRIPTION
**Pull request by Rakshith2597**
Submitting training module for A Two-Stage Multiple Instance Learning Framework for the Detection of Breast Cancer in Mammograms

Paper: [EMBC](https://ieeexplore.ieee.org/document/9176427)
Dataset used for this code repo: [RBIS-DDSM](https://ieee-dataport.org/documents/re-curated-breast-imaging-subset-ddsm-dataset-rbis-ddsm)

This is part of the project MIRIAD: Many Incarnations of Screening of Radiology for High Throughput Disease Screening via Multiple Instance Reinforcement Learning with Adversarial Deep Neural Networks, sponsored by INTEL TECHNOLOGY INDIA PVT. LTD.

Principal Investigators:
Dr Debdoot Sheet (PI), Dr Nirmalya Ghosh (Co-PI)
Department of Electrical Engineering
Indian Institute of Technology Kharagpur

Dr Ramanathan Sethuraman (Co-PI)
Intel Technology India Pvt. Ltd.

### CODE REVIEW

**Comment by Rakshith2597** [comment]
@morkovka1337 This one is ready for review.

File: misc/pytorch_toolkit/mammogram_screening/init_venv.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  echo "$ . venv/bin/activate"
+fi
+
+virtualenv ${venv_dir} -p python3.7 --prompt="(screening)"
+
+
+path_openvino_vars="${INTEL_OPENVINO_DIR:-/opt/intel/openvino}/bin/setupvars.sh"
+if [[ -e "${path_openvino_vars}" ]]; then
+  echo ". ${path_openvino_vars}" >>${venv_dir}/bin/activate
+fi
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/init_venv.sh]: This is no longer needed if you use OpenVINO from the PyPi

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    def inference(self, model, runtype):
+        dc_lst = []
+        to_tensor = transforms.ToTensor()
+        with torch.no_grad():
+            for data in self.dataloader:
+                img = Variable(data['image'].float().to(self.device))
+                mask = Variable(data['mask'].float().to(self.device))
+                if runtype == 'pytorch':
+                    # print(img.shape)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py]: please, delete commented code

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                    out = model(img)
+                    out = torch.sigmoid(out)
+                elif runtype == "onnx":
+                    model = onnxruntime.InferenceSession(self.checkpoint)
+                    ort_inputs = {model.get_inputs()[0].name: to_numpy(img)}
+                    out = model.run(None, ort_inputs)
+                    out = np.array(out)
+                    out = out.squeeze(0).squeeze(1).squeeze(0)
+                    out = to_tensor(out).to(self.device) #.transpose(dim0=1, dim1=0)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py]: please, delete commented code

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/get_config.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import os
+import json
+
+
+def get_config(action, stage='stage1'):
+    """ action: train, test, export or gdrive
+
+    """
+    root_path = os.path.dirname(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
+    config_path = os.path.join(root_path, 'configs')
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/get_config.py]: a bit complicated way to get a config, I would suggest just pass a path to the config

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                dice_coeff = diceCoeff(out, mask, reduce=True)
+                dc_lst.append(dice_coeff.data.cpu().numpy())
+
+        dc_lst = np.array(dc_lst)
+        dice_mean = np.mean(dc_lst)
+        return dice_mean
+
+if __name__ == '__main__':
+
+    config = get_config('inference')
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py]: Suggesting to change the behavior of `get_config` function (see corresponding comment in its definition) and add here an argument parser with parameters: path to a config and run type. Then you won't need commented parts below.

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                    model = onnxruntime.InferenceSession(self.checkpoint)
+                    ort_inputs = {model.get_inputs()[0].name: to_numpy(img)}
+                    out = model.run(None, ort_inputs)
+                    out = np.array(out)
+                    out = out.squeeze(0).squeeze(1).squeeze(0)
+                    out = to_tensor(out).to(self.device) #.transpose(dim0=1, dim1=0)
+
+                else:
+                    out = model.infer(inputs={'input': img})['output']
+                    # print(out.shape)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py]: please, delete commented code

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                    out = out.squeeze(0).squeeze(0)
+                    out = to_tensor(out)#.squeeze(1)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/inference_mass_localization.py]: commented code

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+from ..train_utils.train_function import train_stage1
+from ..train_utils.val_function import val_stage1
+from ..train_utils.get_config import get_config
+from tqdm import tqdm as tq
+
+
+if __name__ == '__main__':
+
+    config = get_config(action='train')
+    epochs = config['epochs']
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py]: same as above about argument parser.
Moreover, consider moving the train procedure into separate function and split the code into logical units. General rule of thumb for splitting function:
1. Function should not be bigger then 1 screen (upon certain exceptions, of course)
2. Function should do only one action according to its name

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/inference_stage2.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+    inference = InferenceStage2(test_loader=tst_loader, checkpoint=config['onnx_checkpoint'], device=device)
+    model = inference.load_model(run_type='onnx')
+    test_acc, auc = inference.inference(model, run_type='onnx', out_nm=out_pred_np)
+    print(test_acc, auc)
+
+    inference = InferenceStage2(test_loader=tst_loader, checkpoint=config['onnx_checkpoint'], device='cpu')
+    model = inference.load_model(run_type='ir')
+    test_acc, auc = inference.inference(model, run_type='ir', out_nm=out_pred_np)
+    print(test_acc, auc)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/inference_stage2.py]: Is it necessary to run all three types of model every time inference script is called? Consider adding run type argument and run only one model per script call.

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/get_config.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import os
+import json
+
+
+def get_config(action, stage='stage1'):
+    """ action: train, test, export or gdrive
+
+    """
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/get_config.py]: looks like `action` annotation does not match the code below

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/loss_functions.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+def bceLoss(mask_pred, mask, pos_weight=28):
+    eps = 1e-15
+    bce = -1.0*pos_weight*mask*torch.log(mask_pred+eps) - (1.0-mask)*torch.log(1.0-mask_pred+eps)
+
+    return torch.mean(bce)
+
+criterion = nn.CrossEntropyLoss()
+
+def ceLoss(pred, target):
+    return criterion(pred, target)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/loss_functions.py]:
```suggestion

def ceLoss(pred, target):
    criterion = nn.CrossEntropyLoss()
    return criterion(pred, target)
```
since this variable is relevant only to a certain function, it should be moved into it

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/step3_get_patches.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+    return (x, y, w, h)
+
+def random_patch(img):
+    while True:
+        # Continue extracting random patches until a patch from the foregorund region is selected.
+        # background has all zero pixel intensities
+        (x, y, w, h) = random_point(img)
+        if np.count_nonzero(img[y:y+h, x:x+w]) > 64*64 - 10:
+            return (x, y, w, h)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/step3_get_patches.py]: Consider adding iterations count to prevent (unlikely, of course, but anyway) endless loop.

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/step3_get_patches.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+def extract_bags(predictions):
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/step3_get_patches.py:44]: Too long function, could you, please, split it into small separate units?

File: misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/train_function.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    for i, data in enumerate(train_loader):
+        img = Variable(data['image'].float().to(device))
+        mask = data['mask'].float().to(device)
+        optimizer.zero_grad()
+        mask_pred = model(img)
+        mask_pred = torch.sigmoid(mask_pred)
+        loss_bce = bceLoss(mask_pred, mask)
+        loss_dice = diceLoss(mask_pred, mask)
+        dice_coeff = diceCoeff(mask_pred, mask, reduce=True)
+        loss = 0.2*loss_bce + 0.8*loss_dice
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/train_function.py]: consider setting loss weights at least as constants in this file, more preferable to have an opportunity to tune them in from the config file (e.g., you can have default values in the model class and read them from the config, if they are specified in it).

**Review by morkovka1337 — CHANGES_REQUESTED** [review]
Overall it is good. A couple of general comments:
1. Please, delete commented code once you do not have need it anymore
2. I think, in the callable scripts (where you have `get_config` function call), it is reasonable to have an argument for the config file path. User can have more than two default configs.

**Comment by nervana-ff** [comment]
Can one of the admins verify this patch?

**Comment by morkovka1337** [comment]
Sorry for delayed reply, will review this week.

**Comment by morkovka1337** [comment]
run tests

**Comment by morkovka1337** [comment]
run misc tests

**Comment by morkovka1337** [comment]
@Rakshith2597, tests are running more than 2 days already. I am afraid, something could be wrong. How long does it take to run tests locally?

**Comment by Rakshith2597** [comment]
> @Rakshith2597, tests are running more than 2 days already. I am afraid, something could be wrong. How long does it take to run tests locally?

@morkovka1337 Locally it takes around 25-30mins (approx).

**Comment by morkovka1337** [comment]
run tests

**Comment by morkovka1337** [comment]
run misc tests

**Comment by morkovka1337** [comment]
Please, fix pylint issues:

```
11:01:31  + python3 tools/run_pylint.py
11:02:03  ERROR:root:pylint failed with code 22
11:02:03  ************* Module mammogram_screening.stage2.step3_get_patches
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/step3_get_patches.py:95:20: W0612: Unused variable 'k' (unused-variable)
11:02:03  ************* Module mammogram_screening.stage2.inference_stage2
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/inference_stage2.py:94:41: C0303: Trailing whitespace (trailing-whitespace)
11:02:03  ************* Module mammogram_screening.stage2.train_stage2
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage2/train_stage2.py:16:13: E1120: No value for argument 'config_path' in function call (no-value-for-parameter)
11:02:03  ************* Module mammogram_screening.stage1.train_stage1
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:16:21: W0621: Redefining name 'config' from outer scope (line 102) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:17:4: W0621: Redefining name 'epochs' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:21:4: W0621: Redefining name 'model_path' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:25:4: W0621: Redefining name 'device' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:34:4: W0621: Redefining name 'train_loader' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:37:4: W0621: Redefining name 'val_loader' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:39:4: W0621: Redefining name 'model' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:42:4: W0621: Redefining name 'optimizer' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:47:4: W0621: Redefining name 'scheduler' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:16: W0621: Redefining name 'epochs' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:24: W0621: Redefining name 'model' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:31: W0621: Redefining name 'train_loader' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:45: W0621: Redefining name 'optimizer' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:56: W0621: Redefining name 'device' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:64: W0621: Redefining name 'val_loader' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:76: W0621: Redefining name 'scheduler' from outer scope (line 103) (redefined-outer-name)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/stage1/train_stage1.py:51:87: W0621: Redefining name 'model_path' from outer scope (line 103) (redefined-outer-name)
11:02:03  ************* Module mammogram_screening.train_utils.models
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/models.py:95:27: E1101: Function 'input' has no 'size' member (no-member)
11:02:03  ************* Module mammogram_screening.train_utils.export
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/export.py:52:13: E1120: No value for argument 'config_path' in function call (no-value-for-parameter)
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/export.py:57:13: E1120: No value for argument 'config_path' in function call (no-value-for-parameter)
11:02:03  ************* Module mammogram_screening.train_utils.transforms
11:02:03  misc/pytorch_toolkit/mammogram_screening/mammogram_screening/train_utils/transforms.py:285:12: W0125: Using a conditional statement with a constant value (using-constant-test)
11:02:03  
11:02:03  -----------------------------------
11:02:03  Your code has been rated at 9.97/10
```

**Comment by Rakshith2597** [comment]
run tests

**Comment by Rakshith2597** [comment]
run misc tests

**Comment by morkovka1337** [comment]
Hi @Rakshith2597

Work on this pull request have to be paused from my side due to the temporal suspension of business of intel in Russia.

**Comment by ryanloney** [comment]
@openvinotoolkit/ote-maintainers can you please review?

**Comment by goodsong81** [comment]
I couldn't have time to figure out how to deal with MISC PRs yet due to urgent release of v0.2.0.
Could you let me know where is the 'run tests / run misc tests' is running?
Link would also be OK.

**Comment by Rakshith2597** [comment]
Any updates on this?

**Comment by goodsong81** [comment]
There seem to be remaining pylint issues.
image

And why don't you add some 'how to run' guide in your ReadMe.md to help others reproduce?
e.g.)
```console
$ ./init_venv.sh
$ ./mammogram_screening/stage1/train_stage1.py ...???
```

Finally, could you run unit tests locally and attach the screen capture here? It would replace CI tests temporarily.

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Please refer to above comments.

**Review by goodsong81 — APPROVED** [review]
LGTM.
